allennlp/tests/data/dataset_readers/masked_language_modeling_test.py

```diff
@@ -32,7 +32,6 @@ def test_text_to_instance_with_bert_tokenizer_and_indexer(self):
             sentence="This is AllenNLP [MASK] token .", targets=["This"]
         )
         assert [t.text for t in instance["tokens"]] == [
-            "[CLS]",
```

Comment:
Ugh, sorry, this actually doesn't work.  For consistency in the library, we actually _need_ the number of tokens here to match the number of tokens that we get out of the `embedder`.  This is so that we can create consistent labelings and have everything work.  Consider, for instance, that I want to label one of these tokens as the correct answer to a question.  I need to pass an index into this token list.  If that token list changes after I create my `Instance` (which happens when you add things in `tokens_to_indices`), then my label is going to be wrong.  I'll have to know about the change, and have code somewhere that changes my label to the right value, meaning that some arbitrary other `Field` has to know about the way that you've done this indexing.  Not good.

Our original pretrained-bert token indexer / embedder handled this by keeping a mapping back to the original tokens, and getting token vectors out again in the embedder.  If we want to add tokens in the `TokenIndexer`, we need to remove them in the `TokenEmbedder` for things to be consistent.  Otherwise things start to fall apart, and the abstractions don't actually work.  And the point of the `pretrained_transformer` indexer and embedder was that we weren't going to do any funny business with the tokens in the embedder, so that means we need to have all of the tokens we need come out of the `tokenizer`.  We _need_ this particular check to stay as it is.  I'm afraid that means we'll have to duplicate some of the logic that's in `transformers`, but I don't know of another way around this.  Can you think of one? 

I just looked through the methods in `transformers`, and the important ones assume that you're adding ids, not strings, so we can't actually use them, unless we do some hacky things inside of `tokens_to_indices` (which maybe we can do...?).

Reply:
I definitely see the point. I ended up with quite a different approach which you can see here: #3453.